drivers/serial/uart_nrfx_uarte.c

```diff
@@ -0,0 +1,607 @@
+		break;
+	case 921600:
+		nrf_baudrate = NRF_UARTE_BAUDRATE_921600;
+		break;
+	case 1000000:
+		nrf_baudrate = NRF_UARTE_BAUDRATE_1000000;
+		break;
+	default:
+		return -EINVAL;
+	}
+
+	nrf_uarte_baudrate_set(uarte, nrf_baudrate);
+
+	return 0;
+}
+
+/**
+ * @brief Poll the device for input.
+ *
+ * @param dev UARTE device struct
+ * @param c Pointer to character
+ *
+ * @return 0 if a character arrived, -1 if the input buffer if empty.
+ */
+static int uarte_nrfx_poll_in(struct device *dev, unsigned char *c)
+{
+	struct uarte_nrfx_data *data = get_dev_data(dev);
+	NRF_UARTE_Type *uarte = get_uarte_instance(dev);
+
+	if (!nrf_uarte_event_check(uarte, NRF_UARTE_EVENT_ENDRX)) {
+		return -1;
+	}
+
+	/* got a character */
+	*c = data->rx_data;
+
+	/* clear the interrupt */
+	nrf_uarte_event_clear(uarte, NRF_UARTE_EVENT_ENDRX);
+	nrf_uarte_task_trigger(uarte, NRF_UARTE_TASK_STARTRX);
+
+	return 0;
+}
+
+/**
+ * @brief Output a character in polled mode.
+ *
+ * @param dev UARTE device struct
+ * @param c Character to send
+ *
+ * @return Sent character
+ */
+static unsigned char uarte_nrfx_poll_out(struct device *dev,
+					 unsigned char c)
+{
+	NRF_UARTE_Type *uarte = get_uarte_instance(dev);
+
+	/* The UART API dictates that poll_out should wait for the transmitter
+	 * to be empty before sending a character. However, without locking,
+	 * this introduces a rare yet possible race condition if the thread is
+	 * preempted between sending the byte and checking for completion.
+
+	 * Because of this race condition, the while loop has to be placed
+	 * after the write to TXD, and we can't wait for an empty transmitter
+	 * before writing. This is a trade-off between losing a byte once in a
+	 * blue moon against hanging up the whole thread permanently
+	 */
+
+	/* reset transmitter ready state */
+	nrf_uarte_event_clear(uarte, NRF_UARTE_EVENT_ENDTX);
+
+	/* send a character */
+	nrf_uarte_tx_buffer_set(uarte, &c, 1);
+	nrf_uarte_task_trigger(uarte, NRF_UARTE_TASK_STARTTX);
+
+	/* Wait for transmitter to be ready */
+	while (!nrf_uarte_event_check(uarte, NRF_UARTE_EVENT_ENDTX)) {
+	}
+
+	/* If there is nothing to send, driver will save an energy
+	 * when TX is stopped.
+	 */
+	nrf_uarte_task_trigger(uarte, NRF_UARTE_TASK_STOPTX);
+
+	return c;
+}
+
+static int uarte_nrfx_err_check(struct device *dev)
+{
+	u32_t error = 0;
+	NRF_UARTE_Type *uarte = get_uarte_instance(dev);
+
+	if (nrf_uarte_event_check(uarte, NRF_UARTE_EVENT_ERROR)) {
+		/* register bitfields maps to the defines in uart.h */
+		error = nrf_uarte_errorsrc_get_and_clear(uarte);
+	}
+
+	return error;
+}
+
+#ifdef CONFIG_UART_INTERRUPT_DRIVEN
+
+/** Interrupt driven FIFO fill function */
+static int uarte_nrfx_fifo_fill(struct device *dev,
+				const u8_t *tx_data,
+				int len)
+{
+	NRF_UARTE_Type *uarte = get_uarte_instance(dev);
+	struct uarte_nrfx_data *data = get_dev_data(dev);
+
+	if (!nrf_uarte_event_check(uarte, NRF_UARTE_EVENT_ENDTX)) {
+		return 0;
```

Comment:
It seems that this check could be skipped. As API doc says:
```
 * @details This function is expected to be called from UART
 * interrupt handler (ISR), if uart_irq_tx_ready() returns true.
 * Result of calling this function not from an ISR is undefined
 * (hardware-dependent). [..]
```

Reply:
fixed